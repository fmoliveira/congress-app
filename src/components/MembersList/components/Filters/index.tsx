import React from "react"
import styled from "styled-components"

import { ControlLabel, Input, SectionTitle } from "../../../_common"

const Container = styled.div`
  margin: 0.5em 1em;
`

const Hint = styled.span`
  margin: 0 1em;
  font-size: 0.8em;
  color: #ae86cb;
`

const Filters = () => (
  <div>
    <SectionTitle>
      Filters
      <Hint>
        Are you looking for a specific senator or representative? Use the
        filters below.
      </Hint>
    </SectionTitle>
    <Container>
      <ControlLabel description="Senator Name">
        <Input type="text" placeholder="Type a name to filter" />
      </ControlLabel>
    </Container>
  </div>
)

export default Filters
