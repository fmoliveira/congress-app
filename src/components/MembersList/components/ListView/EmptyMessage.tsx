import React from "react"
import styled from "styled-components"

import search from "./search.svg"

const MessageWrapper = styled.div`
  margin: 3em;
  text-align: center;
  color: #555555;
`

const ImageWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 150px;
`

const ErrorImage = styled.img`
  width: 80px;
  height: auto;
`

export const EmptyMessage = () => (
  <MessageWrapper>
    <ImageWrapper>
      <ErrorImage src={search} alt="No members found" />
    </ImageWrapper>
    <h2>It's empty here</h2>
    <p>There's no one with the selected filters.</p>
    <p>Please try modifying or removing some filters.</p>
  </MessageWrapper>
)
