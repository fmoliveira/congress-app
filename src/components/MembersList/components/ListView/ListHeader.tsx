import React from "react"

import { lowerCase } from "lodash"
import styled from "styled-components"

import { SmallHeader } from "../../../_common"

const HeaderWrapper = styled.div`
  display: grid;
  grid-template-columns: 2fr repeat(4, 1fr);
  align-items: center;
  margin: 0.75rem;
  padding: 0 0.5rem;
  padding-top: 0.5em;
`

interface IProps {
  chamber: string
}

export const ListHeader = ({ chamber }: IProps) => (
  <HeaderWrapper>
    <SmallHeader>
      {lowerCase(chamber) === "senate" ? "Senator" : "Representative"}
    </SmallHeader>
    <SmallHeader>Party</SmallHeader>
    <SmallHeader>State/District</SmallHeader>
    <SmallHeader>Next Election</SmallHeader>
    <SmallHeader>Social Networks</SmallHeader>
  </HeaderWrapper>
)
