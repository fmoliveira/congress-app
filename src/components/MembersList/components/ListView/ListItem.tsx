import React from "react"
import { withRouter } from "react-router-dom"
import styled from "styled-components"

import { Avatar, PartyName, SocialNetworkLinks } from "../../../_common"

const ItemWrapper = styled.div`
  display: grid;
  grid-template-columns: 2fr repeat(4, 1fr);
  align-items: center;
  margin: 0.75rem;
  padding: 0.5rem;
  background-color: #ffffff;
  color: inherit;
  text-decoration: none;

  &:hover {
    cursor: pointer;
    background-color: #e7e9ef;
  }

  &:active {
    background-color: #e2e5ef;

    & > * {
      transform: translate(1px, 1px);
    }
  }
`

const AvatarWrapper = styled.div`
  display: flex;
  align-items: center;

  * {
    margin-right: 0.5em;
  }
`

const Text = styled.div`
  font-size: 0.8em;
`

interface IProps {
  session: number
  id: string
  firstName: string
  lastName: string
  gender: string
  party: string
  state: string
  nextElection: string
  facebookAccount: string
  twitterAccount: string
  youtubeAccount: string
  history: any
}

export const ListItem = withRouter<any>(
  ({
    session: sessionNumber,
    id: memberId,
    firstName,
    lastName,
    gender,
    party,
    state,
    nextElection,
    twitterAccount,
    facebookAccount,
    youtubeAccount,
    history
  }: IProps) => {
    const goToDetail = () =>
      history.push(`/sessions/${sessionNumber}/members/${memberId}`)
    return (
      <ItemWrapper onClick={goToDetail}>
        <AvatarWrapper>
          <Avatar
            gender={gender}
            facebookAccount={facebookAccount}
            twitterAccount={twitterAccount}
          />
          <Text>
            {firstName} {lastName}
          </Text>
        </AvatarWrapper>
        <Text>
          <PartyName id={party} />
        </Text>
        <Text>{state}</Text>
        <Text>{nextElection}</Text>
        <Text>
          <SocialNetworkLinks
            facebookAccount={facebookAccount}
            twitterAccount={twitterAccount}
            youtubeAccount={youtubeAccount}
          />
        </Text>
      </ItemWrapper>
    )
  }
)
