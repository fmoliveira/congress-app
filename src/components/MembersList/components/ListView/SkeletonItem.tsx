import React from "react"
import styled from "styled-components"

import { CircleSkeleton, TextSkeleton } from "../../../_common"

const ItemWrapper = styled.div`
  display: grid;
  grid-template-columns: 2fr repeat(4, 1fr);
  align-items: center;
  margin: 0.75rem;
  padding: 0.5rem;
  background-color: #ffffff;
`

const AvatarWrapper = styled.div`
  display: flex;
  align-items: center;

  * {
    margin-right: 0.5em;
  }
`

const SocialNetworksWrapper = styled.div`
  * {
    margin: 0 4px;
  }
`

const ColumnSkeleton = ({ children }: { children: React.ReactNode }) => (
  <div>
    <TextSkeleton>{children}</TextSkeleton>
  </div>
)

export const SkeletonItem = () => (
  <ItemWrapper>
    <AvatarWrapper>
      <CircleSkeleton />
      <TextSkeleton>Member Name</TextSkeleton>
    </AvatarWrapper>
    <ColumnSkeleton>Party Name</ColumnSkeleton>
    <ColumnSkeleton>CA</ColumnSkeleton>
    <ColumnSkeleton>Year</ColumnSkeleton>
    <SocialNetworksWrapper>
      <CircleSkeleton />
      <CircleSkeleton />
      <CircleSkeleton />
    </SocialNetworksWrapper>
  </ItemWrapper>
)
