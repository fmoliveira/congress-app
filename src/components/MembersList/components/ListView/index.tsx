import React from "react"

import { RequestStatusType } from "../../../../reducers/requestStatus"

import { ErrorMessage, Paginator } from "../../../_common"
import { IMemberInfo } from "../../types"
import { ListHeader } from "./ListHeader"
import { ListItem } from "./ListItem"
import { SkeletonList } from "./SkeletonList"

interface IProps {
  listMembers: () => void
  members: IMemberInfo[]
  scrollToTop: () => void
  status: RequestStatusType
}

const ListView = ({ listMembers, members, scrollToTop, status }: IProps) => (
  <div>
    {status !== RequestStatusType.Error && <ListHeader />}
    {status === RequestStatusType.Error && <ErrorMessage retry={listMembers} />}
    {status === RequestStatusType.Loading && <SkeletonList />}
    {status === RequestStatusType.Success && (
      <Paginator data={members} onPageChanged={scrollToTop}>
        {i => <ListItem key={i.id} {...i} />}
      </Paginator>
    )}
  </div>
)

export default ListView
