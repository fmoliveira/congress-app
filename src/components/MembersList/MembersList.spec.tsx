import React from "react"
import { cleanup, render } from "react-testing-library"

import Filters from "./components/Filters"
import ListView from "./components/ListView"
import { UnconnectedMembersList as MembersList } from "./index"

jest.mock("./components/Filters", () => jest.fn(() => <div>Filters</div>))
jest.mock("./components/ListView", () => jest.fn(() => <div>ListView</div>))

const setup = (props?: any) => render(<MembersList {...props} />)

describe("MembersList", () => {
  afterEach(() => {
    cleanup()
    jest.clearAllMocks()
  })

  it("Should render without crashing", () => {
    const { container } = setup()
    expect(container.firstChild).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("Should call action to list members", () => {
    const listMembers = jest.fn()
    const props = { listMembers }
    setup(props)

    expect(listMembers).toHaveBeenCalledTimes(1)
    expect(listMembers).toHaveBeenCalledWith(115, "senate")
  })

  it("Should render filters", () => {
    setup()
    expect(Filters).toHaveBeenCalledTimes(1)
  })

  it("Should render list view", () => {
    const props = {
      members: [
        { id: "S100", name: "Edward Thatch" },
        { id: "S101", name: "Anne Bonny" }
      ],
      status: "SUCCESS"
    }
    setup(props)

    expect(ListView).toHaveBeenCalledTimes(1)

    const mockedComponent = ListView as jest.Mock
    const [call] = mockedComponent.mock.calls
    const [passedProps] = call
    expect(passedProps.members).toEqual([
      { id: "S100", name: "Edward Thatch" },
      { id: "S101", name: "Anne Bonny" }
    ])
    expect(passedProps.status).toBe("SUCCESS")
  })

  it("Should allow list view to call list members action", () => {
    const listMembers = jest.fn()
    const props = { listMembers }
    setup(props)

    expect(listMembers).toHaveBeenCalledTimes(1)
    expect(listMembers).toHaveBeenNthCalledWith(1, 115, "senate")

    const mockedComponent = ListView as jest.Mock
    const [call] = mockedComponent.mock.calls
    const [passedProps] = call
    const { listMembers: listMembersProp } = passedProps

    listMembersProp()
    expect(listMembers).toHaveBeenCalledTimes(2)
    expect(listMembers).toHaveBeenNthCalledWith(2, 115, "senate")
  })
})
