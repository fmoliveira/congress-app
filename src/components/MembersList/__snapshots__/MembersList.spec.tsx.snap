// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MembersList Should render without crashing 1`] = `
<div>
  <div>
    <div>
      ListView
    </div>
  </div>
</div>
`;
